### PR TITLE
Add typescript typings

### PR DESCRIPTION
**Pull request by Veetaha**
There are still some TODOs to resolve @octet-stream 
demo
This also means response types maintenance burden tho...

### CODE REVIEW

**Comment by octet-stream** [comment]
Hey, this is actually what I want to add to my library :D But since I’m not familiar with TS, I decided to write JSDoc instead, as you probably noticed. Also, I thought it would be enough for auto completions. At least it works for me when I’m writing this library and tests in VSCode.
Anyway, thank you for the PR! I will take a look tomorrow.

**Comment by Veetaha** [comment]
I recommend TypeScript, it trades a bit more symbols to write (which in fact you already do even more verbosely with JsDoc comments :D) for **consistency** and static **reliability** to the codebase.
I really can't write JS anymore (well I am no longer actively using TS though), but believe me, it saves a lot of overhead (when used with strict configuration).
E.g. the dreaded `"Cannot read property of null | undefined"` is fully solved by TypeScript, because you have to explicitly say `null | T` if you mean that the value can be `null` and the compiler will bite you if you don't handle `null` value when you use it.

You can play with this by creating a sample project 
```
mkdir test-dinky
npm init -y
npm install ~/dev/dinky
npm install -D typescript
npx tsc init
touch main.ts
```
And then try compiling this code in `main.ts`
```ts
import dinky from "dinky.js";

async function main() {
    const res = await dinky({}).comments();
    const reason = res.comments[0].editReason;
    const lowercasedReason = reason.toLocaleLowerCase();
    // if (reason !== null) { // notice that types are contextual, so after this check reason is not nulllable
    //     const lowercasedReason = reason.toLocaleLowerCase();
    // }
}
```
```bash
~/junk/test-dinky $ npx tsc
main.ts:6:30 - error TS2531: Object is possibly 'null'.

6     const lowercasedReason = reason.toLocaleLowerCase();
                               ~~~~~~


Found 1 error.
```

**Comment by octet-stream** [comment]
Man, I have tried TS and I know what it is and how it works. : D By "not familiar" I meant that I haven't learned it, since I don't see any reason for me. But too many people use it, so it would be nice to have typings here in my library.
For me, it gives only one benefit – I have autocompletion in VSCode which means I don't need to open README every time I want to write something (such as tests) for my library : D IntelliSense works with JSDoc just fine so that's why I did improvements for in code documentation.

Is the PR ready for a review or you will add something else first?

**Comment by Veetaha** [comment]
It's pretty much ready, I marked it WIP until there are TODOs you would like to resolve

File: lib/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,801 @@
...
+   * Default options: `{}`
+   */
+  exec<T>(options?: DinkyRequestOptions): Promise<T>
+
+  then: Promise<T>["then"]
+  catch: Promise<T>["catch"]
+}
+
+// TODO: its constructor is marked as public, so why does it have @private annotation then?
+// I guess this doesn't mean it is public to external users, but public inside of this lib or smth...
```
Comment (octet-stream) [inline lib/index.d.ts]: The only public class for external users is Dinky, anything else has `@api private` annotation. This is due to the library's design.

File: lib/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,801 @@
...
+  search(query: (string | string[])[]): Search<EntitiesPage>
+
+  /**
+  * Get one entity by its ID
+  */
+  getById(id: number, options?: DinkyRequestOptions): Promise<Entity>
+}
+
+// TODO: Constructors of these types are not public, but they are described
+// in README, this is weird
```
Comment (octet-stream) [inline lib/index.d.ts]: They are mentioned with description only. This is due to documentation format I've used for my projects : D

File: lib/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,801 @@
...
+  constructor(message: string, response: Response)
+}
+
+interface Request<T> {
+  /**
+   * Sets the page offset.
+   *
+   * Default offset: `1`
+   */
+  page(offset: number): this
```
Comment (octet-stream) [inline lib/index.d.ts]: The `offset` argument must be marked as optional.

File: lib/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,801 @@
...
+export default dinky
+
+export class NetworkError extends Error {
+  response: Response
+  url: string
+  status: number
+  statusText: string
+
+  // TODO: should this one be private?
+  constructor(message: string, response: Response)
```
Comment (octet-stream) [inline lib/index.d.ts]: Yes, this class will be created only in `link` responses.

File: lib/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,801 @@
...
+   */
+  images(): Images
+
+  /**
+   * Creates a `Lists` class that contains a bunch of shortcuts for `Search`
+   */
+  lists(): Lists
+}
+
+export function dinky(options: DinkyRequestOptions): Dinky
```
Comment (octet-stream) [inline lib/index.d.ts]: The options argument is optional everywhere : D

File: lib/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,801 @@
+import type { Response } from "node-fetch"
+export { Response }
```
Comment (octet-stream) [inline lib/index.d.ts]: Move every export to the end of the file please.

File: lib/index.d.ts (hunk 1 above)
Comment (Veetaha) [reply lib/index.d.ts]: Removed this comment in https://github.com/octet-stream/dinky/pull/7/commits/14cfd777a52f1ef319befeb18af6ba8eff751b26
`Entities` interface is left unexported, thus it is a private implementation detail.

File: lib/index.d.ts (hunk 2 above)
Comment (Veetaha) [reply lib/index.d.ts]: I dunno what to say :D...

File: lib/index.d.ts (hunk 3 above)
Comment (Veetaha) [reply lib/index.d.ts]: Good catch! Fixed this in 8ebe26de52d908eeb78f2d425c42cf85b984a687

File: lib/index.d.ts (hunk 4 above)
Comment (Veetaha) [reply lib/index.d.ts]: Made this private and added a comment on the intended use in https://github.com/octet-stream/dinky/commit/14cfd777a52f1ef319befeb18af6ba8eff751b26

File: lib/index.d.ts (hunk 5 above)
Comment (Veetaha) [reply lib/index.d.ts]: Good catch # 2, fixed this in https://github.com/octet-stream/dinky/pull/7/commits/a8fdbf53f65e36070191c51549dbb472b11ff6a2

File: lib/index.d.ts (hunk 6 above)
Comment (Veetaha) [reply lib/index.d.ts]: What do you mean by every export? Do you want to use `export { A, B, C }` to export all the public symbols? I think this is old-fashioned, having an `export` prefixed right at the declaration site is better and is a generally accepted and more modern idiomatic way of ecmascript modules export.
I've reordered the private symbols to appear before the public ones in https://github.com/octet-stream/dinky/pull/7/commits/e55c40ea66ea80087afc8bc472c75c5a5ebdf98d

**Comment by Veetaha** [comment]
@octet-stream this is ready to merge unless you have some ideas for improvement

File: lib/index.d.ts (hunk 6 above)
Comment (octet-stream) [reply lib/index.d.ts]: This is the way how I do prefer to declare modules. Stick with the project's code style, please.
Also, I prefer to omit spaces inside of a curly braces.

**Comment by octet-stream** [comment]
Well, if everything is correct, I can merge this. But I would like to have tests for these typings if necessary.

**Comment by octet-stream** [comment]
I've tried this thing before in my other project, but it didn't worked well for me, when I decided to move different types to different files: https://github.com/SamVerschueren/tsd

Perhaps I did something wrong, doesn't matter.

File: lib/Dinky.d.ts [hunk 7]
```diff
@@ -0,0 +1,821 @@
...
+    /**
+     * Total number of entities that can be retreived by paginating through them.
+     */
+    total: number
+  }
+
+  /**
+   * String in RFC3339 datetime format
+   */
+  export type DateTimeString = string;
```
Comment (octet-stream) [inline lib/Dinky.d.ts]: Btw, dates will be returned as Date, not as a string
See: https://github.com/octet-stream/dinky/blob/bf560fd3f99afd7285b100d721b71f6c50e3ab55/lib/util/castDates.js

Comment (Veetaha) [reply lib/Dinky.d.ts]: Hmm, yeah `Date` type should go there instead

**Comment by octet-stream** [comment]
Hey, if you have no much time to write tests for typings, I can just merge it anyways and add them later.
Either way everything else LGTM.

**Review by octet-stream — APPROVED** [review]

**Comment by Veetaha** [comment]
Regarding last commits [here is the explanation](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-exports)
otherwise it would be a bug to 
```ts
import { Response } from "dinky";
const res = new Response();
```

**Comment by Veetaha** [comment]
> Hey, if you have no much time to write tests for typings, I can just merge it anyways and add them later.
Either way everything else LGTM.

Probably, you a right, I don't have enough time...

**Comment by Veetaha** [comment]
Hmm, I just noticed I didn't export `Response` initially, are you sure we should export it?

**Comment by octet-stream** [comment]
You didn't? Then it was my mistake :D

**Comment by Veetaha** [comment]
Okay, you fix it ;d

**Comment by octet-stream** [comment]
Ok, done :D

**Review by octet-stream — APPROVED** [review]

**Comment by octet-stream** [comment]
Okay, everything seems ok, so I'll merge it.
Thank you so much!

**Comment by octet-stream** [comment]
I will publish this changes in a next release as soon as I add the tests.

**Comment by Veetaha** [comment]
Some day you've got to adopt types and testing them won't be a thing)

**Comment by octet-stream** [comment]
As long as I do small things, I don't really need typings)
But since I've switched to VS Code from Sublime Text I want IntelliSense work correctly, so now I use JSDocs :D

**Pull request merged**